src/neo/Network/P2P/SyncManager.cs

```diff
@@ -0,0 +1,261 @@
+        }
+
+        private void OnReceiveBlockIndex(PersistedBlockIndex persistedIndex)
+        {
+            if (persistedIndex.PersistedIndex <= persistIndex) return;
+            persistIndex = persistedIndex.PersistedIndex;
+            if (persistIndex > lastTaskIndex)
+                lastTaskIndex = persistIndex;
+            if (persistIndex % BlocksPerTask == 0 || persistIndex == highestBlockIndex)
+            {
+                foreach (SyncSession session in sessions.Values.Where(p => p.HasTask))
+                {
+                    for (int i = 0; i < session.Tasks.Count(); i++)
+                    {
+                        if (session.Tasks[i].EndIndex <= persistIndex)
+                        {
+                            session.Tasks.Remove(session.Tasks[i]);
+                            totalTasksCount--;
+                            RequestSync();
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+
+        private void OnTimer()
+        {
+            if (sessions.Count() == 0 || sessions.Values.Where(p => p.HasTask).Count() == 0) return;
+            foreach (SyncSession session in sessions.Values.Where(p => p.HasTask))
+            {
+                for (int i = 0; i < session.Tasks.Count(); i++)
+                {
+                    if (DateTime.UtcNow - session.Tasks[i].Time > SyncTimeout)
+                    {
+                        session.Tasks.Remove(session.Tasks[i]);
+                        session.TimeoutTimes++;
+                        if (!ReSync(session, session.Tasks[i]))
+                            IncrementUncompletedTasks(session.Tasks[i]);
+                    }
+                    if (session.Tasks[i].IndexArray.Cast<bool>().All(p => p == true))
+                    {
+                        totalTasksCount--;
+                    }
+                }
+            }
+            RequestSync();
+        }
+
+        private void OnUpdate(uint lastBlockIndex)
+        {
+            if (!sessions.TryGetValue(Sender, out SyncSession session))
+                return;
+            session.LastBlockIndex = lastBlockIndex;
+        }
+
+        private bool ReSync(SyncSession oldSession, Task task)
+        {
+            Random rand = new Random();
+            SyncSession session = sessions.Values.Where(p => p != oldSession && p.Tasks.Count <= maxTasksPerSession && p.LastBlockIndex >= task.EndIndex).OrderBy(p => p.Tasks.Count).ThenBy(s => rand.Next()).FirstOrDefault();
+            if (session == null)
+                return false;
+            int count = (int)(task.EndIndex - task.StartIndex + 1);
+            session.Tasks.Add(new Task { StartIndex = task.StartIndex, EndIndex = task.EndIndex, IndexArray = new BitArray(count), Time = DateTime.UtcNow });
+            session.RemoteNode.Tell(Message.Create(MessageCommand.GetBlockData, GetBlockDataPayload.Create(task.StartIndex, (ushort)count)));
```

Comment:
I think that we can set in a dictionary wich blocks we want to retrieve in order to use the same message, when we receive a block, check if was requested by the syncmanager or not.